sklearn/base.py

```diff
@@ -990,6 +992,38 @@ def _more_tags(self):
         }
 
 
+class EngineAwareMixin:
+    """Mixin for estimators that use a pluggable engine to do the work"""
+
+    def _get_engine(self, X, y=None, sample_weight=None, reset=False):
+        """Determine the engine for the estimator to use.
+
+        All enabled engine providers are tried in turn, the first one that
+        accepts is selected. The choice of engine is stored and re-used
+        unless `reset=True`, in which case the selection process is started
+        again.
+        """
+        if hasattr(self, "_engine_provider") and not reset:
+            return self._engine_class(self), {
+                "X": X,
+                "y": y,
+                "sample_weight": sample_weight,
+            }
```

Comment:
If the above suggestion is implemented, we should call `engine.validate` here.

Reply:
I thought about that but then left it out. My thinking was that "once the engine has been selected during fit it will be selected forever (until re-fit)". `validate` is used as part of the selection, so why call it again?

However, a reason to call it again is that it performs input conversion. But then we should call it before `fit` and `predict`. Which then makes me wonder what the purpose of `prepare_fit` is (right now it does lots of conversion).

For example in the `KMeans` case `_validate_data` is called from `prepare_fit` and from `prepare_prediction`, but with different arguments (`reset=False` vs `reset=True`). So if we move it to `accept` we need to pass information to `accept` about which context it is being called in (fit or predict). So maybe there should be `accept_fit` and `accept_predict`? But then what is the difference to `prepare_fit` and `prepare_predict`?

Overall it somehow feels like the API isn't "nice" :-/

Reply:
The engine definitely needs to validate and convert the input that the user submits predict methods in some way, IMO it's a core requirement to keep this behavior. Regarding how, I agree that it doesn't speak for itself. It seems that the existing implementation took the most straightforward path to factor identical steps in common engine methods but let's change it if there's another better possible abstraction.

Reply:
> For example in the KMeans case _validate_data is called from prepare_fit and from prepare_prediction, but with different arguments (reset=False vs reset=True). So if we move it to accept we need to pass information to accept about which context it is being called in (fit or predict). So maybe there should be accept_fit and accept_predict?

Maybe we can try to just pass the `reset` argument to the `accept` method for a start? Or alternatively pass a `context="fit"/"predict"/"transform"` string.

Then we can probably remove the `prepare_` methods.

Reply:
After our discussion on Wednesday the PR is now back to using an `accept()` that can't modify the input data. That job is left for the `prepare_*` methods.